node/src/daser.rs

```diff
@@ -110,74 +122,295 @@ where
+            warn!("All peers disconnected");
+            return Ok(());
+        }
+
+        self.rollback().await;
+        self.populate_queue().await;
+
+        loop {
+            if self.sampling_fut.is_terminated() {
+                self.schedule_next_sample_block().await?;
+            }
+
+            select! {
+                _ = self.cancellation_token.cancelled() => {
+                    break;
+                }
+                _ = peer_tracker_info_watcher.changed() => {
+                    if peer_tracker_info_watcher.borrow().num_connected_peers == 0 {
+                        warn!("All peers disconnected");
+                        break;
+                    }
+                }
+                res = &mut self.sampling_fut => {
+                    // Beetswap returns only fatal errors that are not related
+                    // to P2P nor networking.
+                    let (height, accepted) = res?;
+
+                    let status = if accepted {
+                        SamplingStatus::Accepted
+                    } else {
+                        SamplingStatus::Rejected
+                    };
+
+                    self.store
+                        .update_sampling_metadata(height, status, Vec::new())
+                        .await?;
+                },
+                _ = self.store.wait_new_head() => {
+                    self.populate_queue().await;
+                }
+            }
+        }
+
+        self.sampling_fut.terminate();
+        self.queue.clear();
+        self.prev_head = None;
 
         Ok(())
     }
 
-    async fn sample_next_block(&mut self) -> Result<()> {
-        let height = self.store.next_unsampled_height().await?;
-        self.store.wait_height(height).await?;
+    async fn schedule_next_sample_block(&mut self) -> Result<()> {
+        assert!(self.sampling_fut.is_terminated());
+
+        let Some(args) = self.queue.pop_front() else {
+            return Ok(());
+        };
+
+        // Make sure that block is still in the sampling window.
+        if !in_sampling_window(args.time) {
+            // Queue is sorted, so as soon as we reach a block
+            // that is not in the sampling window, it means the
+            // rest wouldn't be either.
+            self.queue.clear();
+            return Ok(());
+        }
+
+        let mut cids = Vec::new();
 
-        let header = self.store.get_by_height(height).await?;
-        let (sampled_cids, accepted) = self.sample_block(&header).await?;
+        // Select random shares to be sampled and generate their CID.
+        for (row, col) in random_indexes(args.square_width, self.max_samples_needed) {
+            let cid = sample_cid(row, col, args.height)?;
+            cids.push(cid);
+        }
 
+        // Update CID list before we start the sampling procedure, otherwise there are
+        // some edge cases that can leak the CIDs and never cleaned from blockstore.
```

Comment:
```suggestion
        // Update the CID list before we start sampling, otherwise it's possible for us to leak CIDs
        // causing associated blocks to never get cleaned from blockstore
```